README.md
# TTPassGen
[![Build Status](https://travis-ci.org/tp7309/TTPassGen.svg?branch=master)](https://travis-ci.org/tp7309/TTPassGen)
[![Coverage Status](https://coveralls.io/repos/github/tp7309/TTPassGen/badge.svg?branch=master)](https://coveralls.io/github/tp7309/TTPassGen?branch=master)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/25f05aa766c34eea9b9692725237e873)](https://www.codacy.com/app/tp7309/TTPassGen?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=tp7309/TTPassGen&amp;utm_campaign=Badge_Grade)

TTPassGen is a highly flexiable and scriptable password dictionary generator base on Python, you can easily use various rules to generate the desired combination of words.

README i18n: [中文说明](https://github.com/tp7309/TTPassGen/blob/master/README_zh_CN.md)

# Features
- generate password use combination、permulation、conditional rules and so on.
- support all characters or words(from wordlist option) that can make up a password, some built-in charset has been provided, such as alphabetical list and numeric list.
- you can specify the order and frequency of each element in the word.
- simple rule format, and easy to use, rule could be defined similar regex's style.
- time-consuming estimates, output size estimates, and real-time progress reports.
- unicode word support by using wordlist option.
- Generation of large amounts of passwords at once, no output size limit.
- it can split output by file size.

# Install
`TTPassGen` can be easily installed using pip:
```
pip install ttpassgen
```
if you are using the windows operating system, you could just use the [release version](https://github.com/tp7309/TTPassGen/releases).

# Requirements
Python 2 (version 2.7 or later), or Python 3 (version 3.3 or later).

# Quick Start
> Switch to the project's `ttpassgen` directory if you want use ttpassgen by downloaded source code.

Example: Generate word list output to *out.dict*, the word format is prefix three digits, only allow 1、2、3, appear 2 or 3 times, followed by letter a or b.
```
ttpassgen -r [123]{2:3}[ab] out.dict
```
Done.

# Options
```
C:\Users\tp730>ttpassgen --help
Usage: ttpassgen [OPTIONS] OUTPUT
Options:
  -m, --mode INTEGER             generation mode:

                                 0 = combination rule mode
                                 [default: 0]
  -d, --dictlist TEXT            read wordlist from the file, multi files
                                 should by seperated by comma.
  -r, --rule TEXT                define word format, $0 means refer first
                                 file in wordlist, some built-in charsets:

                                 ?l = abcdefghijklmnopqrstuvwxyz
                                 ?u = ABCDEFGHIJKLMNOPQRSTUVWXYZ
                                 ?d = 0123456789
                                 ?s = !"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
                                 ?a = ?l?u?d?s
                                 ?q = ]

                                 example: [?dA]{1:2}$0
                                 view *Examples* section for more information.
                                 [default: '']
  -c, --dict_cache INTEGER       each element in 'dictlist' option represents
                                 a dict file path, this option define the
                                 maximum amount of memory(MB) that can be used,
                                 increasing this value when the file is large 
                                 may increase the build speed.  [default: 500]
  -g, --global_repeat_mode TEXT  whether the character is allowd to repeat:

                                 ? = 0 or 1 repetitions
                                 * = 0 or more repetitions
                                 [default: ?]
  -p, --part_size INTEGER        when result data is huge, split package
                                 size(MB) will be applied, 0 is unlimited.
                                 [default: 0]
  -a, --append_mode INTEGER      whether append content to OUTPUT or not.
                                 [default: 0]
  -s, --seperator TEXT           word seperator, by default, each word
                                 occupies one line. special char:

                                 &#160; = one space
                                 [default: ]
  --help                         Show this message and exit.
```
generated password displayed line by line in `OUTPUT`. It is recommended to use Notepad++ to open this file(`utf-8`).

# Examples
**[]**  Used to include charset

## Repeat mode
**[]**  1 repetitions.
`[123] -> 1 2 3`

**[]?** 0 or 1 repetitions
`[123] -> '' 1 2 3`

**[]{m:n:r}**
```
when repeatMode is `?`, [123]{1,2:?} -> 1 2 3 12 13 21 23 31 32
when repeatMode is `*`, [123]{1,2:*} -> 1 2 3 11 12 13 21 22 23 31 32 33
```

**[]{m:n}**
default use `global_repeat_mode` option.

**[]{n} or []{n:r}**
same as `[]{n:n:r}`.

**$no** ref dict file index from `dictlist` option.
```
ttpassgen --dictlist in.dict,in2.dict --rule $0[_]?$1 -s "&#160;" out.dict
when dictlist option defined as `in.dict,in2.dict`,
in.dict content:
word11
word12

in2.dict content:
word21
word22


$0[_]?$1 -> word11word21 word11word22 word11_word21 word11_word22 word12word21 word12word22 word12_word21 word12_word22
```

# Update log
[Update log](https://github.com/tp7309/TTPassGen/blob/master/CHANGES.md)


CHANGES.md
# Update log

## V1.0.2
- add append_mode、seperator、debug_mode options.
- more tests.

## V1.0.1
- Compatible with Python3.
- Supports automatic switching of line seperator output, more compatible with windows notepad.
- Increases the output dictionary size estimate accuracy.

----------

## V1.0.2
- 增加append_mode、seperator、debug_mode选项。
- 修改一些选项错误校验，增加更多测试用例。

## V1.0.1
- 兼容Python3。
- 支持自动切换输出不同系统的换行符。
- 提高输出字典大小预估准确度。

tests/test_TTPassGen.py
# coding: utf-8

from __future__ import print_function
import unittest, os, sys
myPath = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, myPath + '/../')
from ttpassgen import ttpassgen

#test/in.dict generate by: ttpassgen.py -r [123]{3:3} in.dict
class Test_ttpassgen(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        super(Test_ttpassgen, cls).setUpClass()

        def lc_func():
            if not os.path.exists('testout.dict'): return 0
            with open('testout.dict', 'r') as f:
                return len(f.readlines())
        global lc
        lc = lc_func

        def go_func(rule, dictlist=os.path.join(myPath, 'in.dict'), mode=0, partSize=0, diskCache=500, repeatMode='?', seperator=None, debugMode=1):
            try:
                ttpassgen.cli.main(['-m', mode, '-d', dictlist, '-r', rule, '-g', repeatMode, '-c',
                    diskCache, '-p', partSize, '-s', seperator, '--debug_mode', debugMode, 'testout.dict'])
            except(SystemExit):
                pass
            return lc()
        global go
        go = go_func


    @classmethod
    def tearDownClass(cls):
        super(Test_ttpassgen, cls).tearDownClass()
        lc = None
        go = None
        if os.path.exists('testout.dict'): os.remove('testout.dict')
        if os.path.exists('testout.dict.1'): os.remove('testout.dict.1')
        if os.path.exists('testout.dict.2'): os.remove('testout.dict.2')
        if os.path.exists('testout.dict.3'): os.remove('testout.dict.3')
        

    def test_invalid_options(self):
        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('$0', mode=233), 0)

        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go(dictlist=None, rule=None), 0)

        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('[?d]', repeatMode="cc9"), 0)

        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('$0word233', dictlist=''), 0)

        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('$99[hello]'), 0)

        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('$0', dictlist='not_exist.dict'), 0)


    def test_not_exist_output_file(self):
        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('$0'), 6)
    
    def test_dict_copy_rule(self):
        self.assertEquals(go('$0'), 6)


    def test_charset_rule_no_range(self):
        self.assertEquals(go('[abc]'), 3)


    def test_charset_rule(self):
        self.assertEquals(go('[abc]?'), 4)


    def test_charset_rule_with_range(self):
        self.assertEquals(go('[abc]{2:3}'), 12)

    
    def test_charset_rule_with_global_repeat_mode(self):
        self.assertEquals(go('[abc]{2:3}', repeatMode='*'), 36)


    def test_mask_charset_rule(self):
        self.assertEquals(go('[?d]?'), 11)


    def test_mask_charset_rule_with_range(self):
        self.assertEquals(go('[a?dA]{1:2}'), 144)
        self.assertEquals(go('[?d]{2:2}'), 90)
        self.assertEquals(go('[?d]{2}'), 90)
    

    def test_dict_mark_charset_rule(self):
        self.assertEquals(go('$0[abc]?'), 24)


    def test_word_seperator(self):
        self.assertEquals(go('$0[abc]?', seperator='&#160;'), 1)
        self.assertEquals(go('$0[abc]?', seperator='-------------------------\n'), 24)
    

    def test_multi_dict_mark_charset_rule(self):
        inDict = os.path.join(myPath, 'in.dict')
        in2Dict = os.path.join(myPath, 'in2.dict')
        with open(in2Dict, 'wb') as f:
            content = ['q00', 'q01']
            f.write(('\n'.join(content)).encode('utf-8'))
        self.assertEquals(go('$1$0[abc]?', dictlist="%s,%s"%(inDict, in2Dict)), 48)
        if os.path.exists(in2Dict): os.remove(in2Dict)


    def test_part_size_with_complex_rule(self):
        if os.path.exists('testout.dict.1'): os.remove('testout.dict.1')
        if os.path.exists('testout.dict.2'): os.remove('testout.dict.2')
        if os.path.exists('testout.dict.3'): os.remove('testout.dict.3')

        go('[?d]{1:4:*}$0[?q]$[0123]', partSize=1, debugMode=1)
        totalLine = 0
        with open('testout.dict.1', 'r') as f:
            totalLine += len(f.readlines())
        with open('testout.dict.2', 'r') as f:
            totalLine += len(f.readlines())
        with open('testout.dict.3', 'r') as f:
            totalLine += len(f.readlines())

        #actual value: 1024 * 1, why 24 difference? I like do it>_>
        self.assertTrue(1000 <= os.path.getsize('testout.dict.1') / 1024 <= 1048)
        if os.path.exists('testout.dict.1'): os.remove('testout.dict.1')
        if os.path.exists('testout.dict.2'): os.remove('testout.dict.2')
        if os.path.exists('testout.dict.3'): os.remove('testout.dict.3')
        self.assertEquals(totalLine, 266640)


    def test_multiprocessing_complex_rule(self):
        self.assertEquals(go('[789]{0:3:*}$0[?q]$0', debugMode=0), 1440)
        self.assertEquals(go('[789]{0:3:*}$0[?q]{1:?}$0', debugMode=0), 1440)

    
    def test_dick_cache(self):
        self.assertEquals(go('$0[abc]?', diskCache=0), 24)

    
    def test_part_size(self):
        if os.path.exists('testout.dict.1'): os.remove('testout.dict.1')
        if os.path.exists('testout.dict.2'): os.remove('testout.dict.2')
        if os.path.exists('testout.dict.3'): os.remove('testout.dict.3')

        go('[?l]{1:4}', partSize=1)
        totalLine = 0
        with open('testout.dict.1', 'r') as f:
            totalLine += len(f.readlines())
        with open('testout.dict.2', 'r') as f:
            totalLine += len(f.readlines())
        if len(os.linesep) > 1:
            with open('testout.dict.3', 'r') as f:
                totalLine += len(f.readlines())

        #actual value: 1024 * 1, why 24 difference? I like do it>_>
        self.assertTrue(1000 <= os.path.getsize('testout.dict.1') / 1024 <= 1048)
        if os.path.exists('testout.dict.1'): os.remove('testout.dict.1')
        if os.path.exists('testout.dict.2'): os.remove('testout.dict.2')
        if os.path.exists('testout.dict.3'): os.remove('testout.dict.3')
        self.assertEquals(totalLine, 375076)

    
    def test_wrong_length_in_rule(self):
        if os.path.exists('testout.dict'): os.remove('testout.dict')
        self.assertEquals(go('[789]{0:-3}'), 0)


if __name__ == '__main__':
    unittest.main()

ttpassgen/ttpassgen.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

from __future__ import print_function
import sys
import click
import re
import itertools
import time
import os
import math, functools
from multiprocessing import Array
import multiprocessing
import threading
from tqdm import tqdm
from collections import OrderedDict


# Module multiprocessing start: organized differently in Python 3.4+
try:
    # Python 3.4+
    if sys.platform.startswith('win'):
        import multiprocessing.popen_spawn_win32 as forking
    else:
        import multiprocessing.popen_fork as forking
except ImportError:
    import multiprocessing.forking as forking

if sys.platform.startswith('win'):
    # First define a modified version of Popen.
    class _Popen(forking.Popen):
        def __init__(self, *args, **kw):
            if hasattr(sys, 'frozen'):
                # We have to set original _MEIPASS2 value from sys._MEIPASS
                # to get --onefile mode working.
                os.putenv('_MEIPASS2', sys._MEIPASS)
            try:
                super(_Popen, self).__init__(*args, **kw)
            finally:
                if hasattr(sys, 'frozen'):
                    # On some platforms (e.g. AIX) 'os.unsetenv()' is not
                    # available. In those cases we cannot delete the variable
                    # but only set it to the empty string. The bootloader
                    # can handle this case.
                    if hasattr(os, 'unsetenv'):
                        os.unsetenv('_MEIPASS2')
                    else:
                        os.putenv('_MEIPASS2', '')

    # Second override 'Popen' class with our modified version.
    forking.Popen = _Popen

class Process(multiprocessing.Process):
    pass
#Module multiprocessing end

_modes = OrderedDict([
    (0, 'combination rule mode')
])

_built_in_charset = OrderedDict([
    ("?l", "abcdefghijklmnopqrstuvwxyz"),
    ("?u", "ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
    ("?d", "0123456789"),
    ("?s", "!\"#$%&'()*+,-./:;<=>?@[\]^_`{|}~"),
    ("?a", "?l?u?d?s"),
    ("?q", "]")
])

_repeat_modes = OrderedDict([
    ("?", "0 or 1 repetitions"),
    ("*", "0 or more repetitions")
])

_special_seperators_name = OrderedDict([
    ("&#160;", "one space")
])

_special_seperators = OrderedDict([
    ("&#160;", " ")
])


_part_dict_name_format = '%s.%d'


#Rule class start
class CharsetRule(object):
    def __init__(self, minLength, maxLength, charset, repeatMode):
        self.minLength = minLength;
        self.maxLength = maxLength;
        self.charset = charset;
        self.repeatMode = repeatMode;


class DictRule(object):
    def __init__(self, order, dictPath):
        self.order = order;
        self.dictPath = dictPath;
#Rule class end

class WordProductor(object):
    def __init__(self, countList, sizeList, productors):
        self.countList = countList
        self.sizeList = sizeList
        self.productors = productors

    @classmethod
    def prod(cls, iterable):
        p= 1
        for n in iterable: p *= n
        return p

    def totalCount(self):
        return self.prod(self.countList)
    
    def totalSize(self, sep=os.linesep):
        total = 0
        tCount = self.totalCount()
        for i, size in enumerate(self.sizeList):
            total += size * tCount / self.countList[i]
        total += tCount * len(sep)
        return total


def prettySize(size_bytes):
    if size_bytes == 0:
       return "0 Bytes"
    size_name = ("Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
    i = int(math.floor(math.log(size_bytes, 1024)))
    p = math.pow(1024, i)
    s = round(size_bytes / p, 2)
    return "%s %s" % (s, size_name[i])

def formatDict(d):
    return '\n'.join(['%s = %s' % (key, value) for (key, value) in d.items()])


def getExpandedCharset(charset):
    expandedCharset = charset
    for (key, value) in _built_in_charset.items():
        expandedCharset = expandedCharset.replace(key, value)
    return expandedCharset


def echoCommonTips(varName):
    click.echo(
        "%s is invalid, , try use 'python TDictGen.py --help' for get more information." % (varName))


def getCharsetRuleResultDataSize(rule):
    size = count = float(0)
    if rule.repeatMode == '?':
        for wordLength in range(rule.minLength, rule.maxLength + 1):
            subSum = 1
            for i in range(len(rule.charset) - wordLength + 1, len(rule.charset) + 1):
                subSum *= i
            count += subSum
            size += subSum * wordLength
    else:
        for wordLength in range(rule.minLength, rule.maxLength + 1):
            count += math.pow(len(rule.charset), wordLength)
            size += count * wordLength
    return count, size


def getDictRuleResultDataSize(rule):
    sumLines = 0
    sepLen = 1
    with open(rule.dictPath, 'r') as f:
        bufferSize = 1024 * 4
        readFunc = f.read #loop optimization
        chunk = readFunc(bufferSize)
        if '\r\n' in chunk: sepLen = 2
        while chunk:
            sumLines += chunk.count('\n')
            chunk = readFunc(bufferSize)
    linSeperatorCount = sepLen * sumLines
    return sumLines, (os.path.getsize(rule.dictPath) - linSeperatorCount)


def charsetWordProductorWrapper(func):
    @functools.wraps(func)
    def wrapper(*args, **kw):
        f = func(*args, **kw)
        for item in f:
            yield ''.join(item) #loop optimization
    return wrapper

@charsetWordProductorWrapper
def charsetWordProductor(repeatMode, expandedCharset, length):
    if repeatMode == '?':
        return itertools.permutations(expandedCharset, r=length)
    else:
        return itertools.product(expandedCharset, repeat=length)

def largeDictWordProductor(rule):
    #NTFS default block size is 4096 bytes or bigger, try align it.
    with open(rule.dictPath, 'r', buffering=1024 * 4) as f:
        for line in f:
            yield line.strip()


def normalDictWordProductor(rule):
    with open(rule.dictPath, 'r') as f:
        return f.read().splitlines()
    

def generateCombinationDict(mode, dictList, rule, dictCache, globalRepeatMode, partSize, 
    appendMode, seperator, debug_mode, output):
    if not dictList and not rule:
        click.echo("dictlist and rule option must have at least one value!")
        return
    rules = extractRules(dictList, rule, globalRepeatMode)
    if not rules:
        echoCommonTips("rule")
        return
    
    dictFiles = []
    if dictList:
        dictFiles = re.split(r',\s*', dictList) if dictList else []
        for f in dictFiles:
            if not os.path.exists(f):
                echoCommonTips("dictlist")
                return
            
    if not globalRepeatMode or not re.match(r"\?|\*", globalRepeatMode):
        echoCommonTips("global_repeat_mode")
        return

    print(("mode: %s, global_repeat_mode: %s, part_size: %s, dictlist: %s, rule: %s")
               % (_modes[mode], globalRepeatMode, prettySize(partSize * 1024 * 1024), dictFiles, rule))
    result = Array('i', [0, 0, 0, 0], lock=False) #[countDone, wordCount, progress, finishFlag]
    if not debug_mode:
        worker = Process(target=productCombinationWords, args=(result, rules, dictCache, partSize, appendMode, seperator, output))
    else:
        worker = threading.Thread(target=productCombinationWords, args=(result, rules, dictCache, partSize, appendMode, seperator, output))
    worker.start()
    while not result[0]: time.sleep(0.05)
    pbar = tqdm(total=result[1], unit=' word')
    progress = 0
    while progress < pbar.total:
        time.sleep(0.1)
        if result[3]: break #exit
        delta = result[2] - progress
        progress += delta
        pbar.update(delta)
    worker.join()
    if (progress < pbar.total): pbar.update(pbar.total - progress)  #avoid stay at 99%
    pbar.close()
    click.echo("generate dict complete.")


def extractRules(dictList, rule, globalRepeatMode):
    splitedDict = re.split(r',\s*', dictList) if dictList else []
    dictCount = len(splitedDict)
    reCharset = r"(\[([^\]]+?)\](\?|(\{-?\d+:-?\d+(:[\?\*])?\})|(\{-?\d+(:[\?\*])?\}))?)"
    reDict = r"(\$(\d{1,%s}))"%(dictCount if dictCount > 0 else 1)
    reRule = r"%s|%s"%(reCharset, reDict)
    rules = []
    matchesLength = 0
    matches = re.findall(reRule, rule)
    try:
        for match in matches:
            match = list(filter(len, match)) #may have some empty element
            matchesLength += len(match[0])
            if re.match(reDict, match[0]):
                order = int(match[1])
                dictRule = DictRule(order, splitedDict[order])
                rules.append(dictRule)
            else:
                minLength = 0
                maxLength = 1
                expandedCharset = getExpandedCharset(match[1])
                repeatMode = globalRepeatMode
                if len(match) > 2:
                    lenInfo = match[2][1:-1].split(':')
                    if len(lenInfo) >= 2:
                        if re.match('-?\d+', lenInfo[1]):  #[]{minLength:maxLength:repeatMode}
                            minLength = int(lenInfo[0])
                            maxLength = int(lenInfo[1])
                            if (len(lenInfo) > 2):
                                repeatMode = lenInfo[2]
                        else:
                            minLength = maxLength = int(lenInfo[0])
                            repeatMode = lenInfo[1]
                    elif match[2][0] == '{':
                        minLength = maxLength = int(lenInfo[0])  #[]{n}
                    else:
                        repeatMode = lenInfo[0]    #?
                else:
                    minLength = maxLength = 1
                if minLength < 0 or maxLength < 0 or minLength > maxLength or len(expandedCharset) < maxLength:
                    return None
                charsetRule = CharsetRule(minLength, maxLength, expandedCharset, repeatMode)
                rules.append(charsetRule)
    except IndexError:
        return None
    if matchesLength <= 0: return None
    return rules


def generateWordProductor(rules, dictCacheLimit):
    wordCountList = []
    wordSizeList = []
    wordProductors = []
    dictCaches = {}
    restCache = dictCacheLimit
    for ruleObj in rules:
        if isinstance(ruleObj, CharsetRule):
            wordCount, wordSize = getCharsetRuleResultDataSize(ruleObj)
            wordCountList.append(wordCount)
            wordSizeList.append(wordSize)
            p = []
            for length in range(ruleObj.minLength, ruleObj.maxLength + 1):
                p.append(charsetWordProductor(ruleObj.repeatMode, ruleObj.charset, length))
            wordProductors.append(itertools.chain(*p) if len(p) > 1 else p[0])
        else:
            if ruleObj.dictPath in dictCaches:
                wordProductors.append(dictCaches[ruleObj.dictPath])
            else:
                fileSize = os.path.getsize(ruleObj.dictPath)
                if fileSize > restCache:
                    wordProductors.append(largeDictWordProductor(ruleObj))
                else:
                    dictCaches[ruleObj.dictPath] = normalDictWordProductor(ruleObj)
                    wordProductors.append(dictCaches[ruleObj.dictPath])
                    restCache -= fileSize
            wordCount, wordSize = getDictRuleResultDataSize(ruleObj)
            wordCountList.append(wordCount)
            wordSizeList.append(wordSize)
    return WordProductor(wordCountList, wordSizeList, wordProductors)


def productCombinationWords(result, rules, dictCacheLimit, partSize, appendMode, seperator, output):
    productor = generateWordProductor(rules, dictCacheLimit)
    result[1] = int(productor.totalCount())
    result[0] = 1
    estimatedSize = prettySize(productor.totalSize(sep=seperator))
    print(("estimated size: %s, generate dict...")%(estimatedSize))

    if not os.path.exists(os.path.abspath(os.path.join(output, os.path.pardir))):
        os.makedirs(os.path.abspath(os.path.join(output, os.path.pardir)))
    realPartSize = partSize * 1024 * 1024
    partIndex = 1
    currSize = 0
    firstOutputFileName = output
    if realPartSize: firstOutputFileName = _part_dict_name_format%(output, partIndex)

    wordSeperator = os.linesep
    if seperator:
        wordSeperator = _special_seperators[seperator] if seperator in _special_seperators else seperator

    lineSeperatorLength = len(wordSeperator)
    fileMode = 'ab' if appendMode else 'wb'

    progress = 0
    def progressMonitor():
        while not result[3]:
            result[2] = progress
            time.sleep(0.1)
    threading.Thread(target=progressMonitor).start()

    try:
        p = itertools.product(*productor.productors) if len(productor.productors) > 1 else productor.productors[0]
        f = open(firstOutputFileName, fileMode, buffering=1024 * 4)

        # wrap f.write('content') as func will reduce the generation speed, so remain if...elif...
        if sys.version_info > (3, 0): #there will be a minimum of 10% performance improvement.
            if realPartSize:  # avoid condition code cost.
                if len(productor.productors) > 1:  # complex rule, more join cost.
                    for w in p:
                        content = (''.join(w) + wordSeperator).encode('utf-8')
                        f.write(content)
                        progress += 1
                        lineLength = len(content)
                        currSize += lineLength
                        if currSize > realPartSize:
                            currSize = lineLength
                            f.close()
                            partIndex += 1
                            f = open(_part_dict_name_format % (
                                output, partIndex), fileMode, buffering=1024 * 4)
                else:
                    for w in p:
                        content = (w + wordSeperator).encode('utf-8')
                        f.write(content)
                        progress += 1
                        lineLength = len(content)
                        currSize += lineLength
                        if currSize > realPartSize:
                            currSize = lineLength
                            f.close()
                            partIndex += 1
                            f = open(_part_dict_name_format % (
                                output, partIndex), fileMode, buffering=1024 * 4)
            else:
                if len(productor.productors) > 1:  # complex rule, more join cost.
                    for w in p:
                        f.write((''.join(w) + wordSeperator).encode('utf-8'))
                        progress += 1
                else:
                    for w in p:
                        f.write((w + wordSeperator).encode('utf-8'))
                        progress += 1
        else:
            if realPartSize:  # avoid condition code cost.
                if len(productor.productors) > 1:  # complex rule, more join cost.
                    for w in p:
                        content = ''.join(w) + wordSeperator
                        f.write(content)
                        progress += 1
                        lineLength = len(content)
                        currSize += lineLength
                        if currSize > realPartSize:
                            currSize = lineLength
                            f.close()
                            partIndex += 1
                            f = open(_part_dict_name_format % (
                                output, partIndex), fileMode, buffering=1024 * 4)
                else:
                    for w in p:
                        content = w + wordSeperator
                        f.write(content)
                        progress += 1
                        lineLength = len(content)
                        currSize += lineLength
                        if currSize > realPartSize:
                            currSize = lineLength
                            f.close()
                            partIndex += 1
                            f = open(_part_dict_name_format % (
                                output, partIndex), fileMode, buffering=1024 * 4)
            else:
                if len(productor.productors) > 1:  # complex rule, more join cost.
                    for w in p:
                        f.write(''.join(w) + wordSeperator)
                        progress += 1
                else:
                    for w in p:
                        f.write(w + wordSeperator)
                        progress += 1
    finally:
        f.close()
    result[3] = 1

@click.command()
@click.option("--mode", "-m", show_default=True, default=0, type=click.INT, help="generation mode:\n\n" + formatDict(_modes))
@click.option("--dictlist", "-d", type=click.STRING,
              help="read wordlist from the file, multi files should by seperated by comma.")
@click.option("--rule", "-r", type=click.STRING, show_default=True, default="",
              help="define word format, $0 means refer first file in wordlist, some built-in charsets:\n\n"
                + formatDict(_built_in_charset) + "\n\nexample: [?dA]{1:2}$0\nview documentation for more information.")
@click.option("--dict_cache", "-c", type=click.INT, show_default=True, default=500,
              help="each element in 'dictlist' option represents a dict file path, this option define"
                + " the maximum amount of memory(MB) that can be used when reading their contents,"
                + "increasing this value when the file is large may increase the build speed.")
@click.option("--global_repeat_mode", "-g", show_default=True, default='?', type=click.STRING,
              help="whether the character is allowd to repeat:\n\n" + formatDict(_repeat_modes))
@click.option("--part_size", "-p", type=click.INT, default=0, show_default=True,
              help="when result data is huge, split package size(MB) will be applied, 0 is unlimited.")
@click.option("--append_mode", "-a", type=click.INT, default=0, show_default=True,
              help="whether append content to OUTPUT or not.")
@click.option("--seperator", "-s", type=click.STRING, default='', show_default=True,
              help="word seperator, by default, each word occupies one line. special char:\n\n" + formatDict(_special_seperators_name))
@click.option("--debug_mode", type=click.INT, default=0, show_default=True,
              help="set 1 for enter debug mode.")
@click.argument("output", type=click.Path())
def cli(mode, dictlist, rule, dict_cache, global_repeat_mode, part_size, append_mode, seperator, debug_mode, output):
    # On Windows calling this function is necessary.
    multiprocessing.freeze_support()
    
    if mode in _modes:
        generateCombinationDict(mode, dictlist, rule, dict_cache, global_repeat_mode, part_size, append_mode, seperator, debug_mode, output)
    else:
        click.echo(
            "unknown mode, try use 'python TDictGen.py --help' for get more information.")


if __name__ == "__main__":
    cli()
    # cli.main(['-d', '../tests/in.dict', '-r', '[?l]{5}', 'out.dict'])
